### PR TITLE
Add support for more protocol op

### PR DESCRIPTION
**Pull request by jiegec**
Note: the code is not carefully verified and might contain bugs.

### CODE REVIEW

File: src/proto.rs [hunk 1]
```diff
@@ -675,6 +734,45 @@ impl TryFrom<StructureTag> for LdapFilter {
...
+
+                let f = inner
+                    .pop()
+                    .and_then(|t| t.match_class(TagClass::Universal))
+                    .and_then(|t| t.match_id(Types::Sequence as u64))
+                    .and_then(|t| t.expect_constructed())
+                    .and_then(|bv| {
+                        let mut res: Vec<LdapSubstringFilter> = vec![];
+                        for StructureTag { class, id, payload } in bv.iter() {
+                            res.push(match (id, payload) {
```
Comment (Firstyear) [inline src/proto.rs:757]: I don't think this quite correct. A substring filter can have 0 or 1 Initial, and 0 or 1 Final, and 0 to infinite Any. So I think that the content of "Substring(String, Vec<LdapSubstringFilter>)," should change to reflect this, perhaps it should have a structure for LdapSubstringFilter, which has:

```
struct LdapSubstringFilter {
    initial: Option<String>,
    any: Vec<String>,
    final: Option<String>,
}
```

And this parser should reject any substring filter where initial/final occur more than once.

**Review by Firstyear — CHANGES_REQUESTED** [review]
It would be good to have some basic reflection tests in lib.rs (see the use of the do_test macro). Could also be good to wire these into the example and give them a test, but in general this looks really well done, great work. Thanks for the contribution!

File: src/proto.rs (hunk 1 above)
Comment (jiegec) [reply src/proto.rs]: Done.

**Comment by Firstyear** [comment]
I think this looks really good, thank you very much!

**Pull request merged**